server/app/repository/ProgramRepository.java

```diff
@@ -123,6 +126,35 @@ public ImmutableSet<String> getAllProgramNames() {
     return names.build();
   }
 
+  /**
+   * Gets the program definition that contains the related question data from the cache (if
+   * enabled).
+   */
+  public Optional<ProgramDefinition> getFullProgramDefinitionFromCache(ProgramModel program) {
+    if (settingsManifest.getQuestionCacheEnabled()) {
+      return programDefCache.get(String.valueOf(program.id));
+    }
+    return Optional.empty();
+  }
+
+  public Optional<ProgramDefinition> getFullProgramDefinitionFromCache(long programId) {
+    if (settingsManifest.getQuestionCacheEnabled()) {
+      return programDefCache.get(String.valueOf(programId));
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * Sets the program definition that contains the related question data in the cache (if enabled).
+   *
+   * <p>Draft program definition data must not be set in the cache.
```

Comment:
question (blocking): Is there a way to assert that the `programDefinition` isn't a draft before saving it in the cache?

Reply:
Yeah this is a good question. The reason I didn't do it is because we only call this function if it isn't a draft and finding out if a program is a draft requires an extra lookup to get the active version and then get the versions for this program ID, so I thought it may not be worth the extra call. But if we think this may be used in situations it shouldn't it may be worth doing this. This is what the extra code would look like. Let me know what you think!

```
    VersionModel activeVersion = versionRepository.getActiveVersion();
    VersionModel maxVersionForProgram =
        programRepository.getVersionsForProgram(program).stream()
            .max(Comparator.comparingLong(p -> p.id))
            .orElseThrow();
    // If the max version is less than or equal to the active version, it is not a draft
    if (maxVersionForProgram.id <= activeVersion.id) {
```

Reply:
I think it's probably still worth it, just because it's easy for someone in the future to call this method without remembering to read the documentation.

This method will only be called when an admin does a publishing action, right? So the extra call shouldn't happen that frequently and that seems worth it for the extra safety guarantee. If this method could also be called on the applicant side then I agree with you it wouldn't be worth it.

Reply:
Ah got it, thanks for the clarification that it could happen on the applicant side.